reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java

```diff
@@ -91,17 +91,17 @@ public void subscribe(Subscriber<? super T> s) {
 		Objects.requireNonNull(s, "s");
 
 		ColdTestPublisherSubscription<T> p = new ColdTestPublisherSubscription<>(s, this);
-		s.onSubscribe(p); // will trigger drain() via request()
-
 		ColdTestPublisher.SUBSCRIBED_COUNT.incrementAndGet(this);
 
-		if (p.cancelled) {
-			return;
-		}
-
 		if (!add(p)) {
-			p.drain();
+			s.onSubscribe(p); // will trigger drain() via request()
+			if (p.cancelled) {
+				return;
+			}
+			p.drain(); // ensures that empty source terminal signal is propagated without waiting for a request from the subscriber
 		}
+
+		s.onSubscribe(p); // will trigger drain() via request()
```

Comment:
we have potentially 2 onSubscribe calls now

Reply:
I added a return statement in the if statement, so we should not have it